---
title: "R_assignment_testing"
author: "David Hall"
date: "2023-03-12"
output: html_document
---

# R Assignment

load necessary R packages:
```{r packages}
library(tidyverse)
library(tidyr)
```

load data files `fang_et_al_genotypes.txt` and `snp_position.txt`:
```{r data}
fang_et_al_genotypes <- 
  read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2023/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt")

snp_position <- 
  read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2023/main/assignments/UNIX_Assignment/snp_position.txt")
```

## Data inspection
### Properties of `fang_et_al_genotypes`
### Properties of `snp_position`
## Data processing

### Maize data

Extract maize data, transpose, and convert SNP_ID from row names to a column:
```{r sort_transpose}
maize_genotypes <- filter(fang_et_al_genotypes, Group=='ZMMIL'|Group =='ZMMLR'|Group =='ZMMMR') %>% t() %>% as.data.frame() %>% rownames_to_column()
names(maize_genotypes)[1] <- "SNP_ID"
```

Join SNP_ID, chromosome, and position fields of `snp_position` with genotype data:
```{r join}
maize_genotypes_positions <- c(snp_position["SNP_ID"],snp_position["Chromosome"],snp_position["Position"]) %>% as.data.frame() %>% left_join(maize_genotypes, by = "SNP_ID")
```

sort by chromosome:
```{r chromosomes}
for (i in 1:10) {
  
  #filter by chromosome and arrange by increasing position value
  temp_object1 <- filter(maize_genotypes_positions, Chromosome == i & Position != 'multiple' & Position != 'unknown') %>% arrange(as.numeric(Position))
  
  #arrange by decreasing position value and replace '?' with '-'
  temp_object2 <- temp_object1 %>% arrange(desc(as.numeric(Position))) %>% mutate_if(is.character, str_replace_all, pattern = '[?]', replacement = '-')
  
  #assign stored data to object named by chromosome
  assign(paste0('maize_incr_chr_', i), temp_object1)
  assign(paste0('maize_decr_chr_', i), temp_object2)
}
```


### Teosinte data

Extract teosinte data, transpose, and convert SNP_ID from row names to a column:
```{r sort_transpose}
teosinte_genotypes <- filter(fang_et_al_genotypes, Group =='ZMPBA'|Group =='ZMPIL'|Group =='ZMPJA') %>% t() %>% as.data.frame() %>% rownames_to_column()
names(teosinte_genotypes)[1] <- "SNP_ID"
```

Join chromosome and position fields of `snp_position` with genotype data:
```{r join}
teosinte_genotypes_positions <- c(snp_position["SNP_ID"],snp_position["Chromosome"],snp_position["Position"]) %>% as.data.frame() %>% left_join(teosinte_genotypes, by = "SNP_ID")
```

sort by chromosome:
```{r chromosomes}
for (i in 1:10) {
  
  #filter by chromosome and arrange by increasing position value
  temp_object1 <- filter(teosinte_genotypes_positions, Chromosome == i & Position != 'multiple' & Position != 'unknown') %>% arrange(as.numeric(Position))
  #arrange by decreasing position value and replace '?' with '-'
  temp_object2 <- temp_object1 %>% arrange(desc(as.numeric(Position))) %>% mutate_if(is.character, str_replace_all, pattern = '[?]', replacement = '-')
  
  #assign stored data to object named by chromosome
  assign(paste0('teosinte_incr_chr_', i), temp_object1)
  assign(paste0('teosinte_decr_chr_', i), temp_object2)
}
```

## Visualization

### SNPs per chromosome

Distribution of SNPs on each chromosome
```{r}

ggplot(data = maize_genotypes_positions) + geom_point(aes(x = as.numeric(Position), y = Chromosome))

ggplot(data = teosinte_genotypes_positions) + geom_point(aes(x = as.numeric(Position), y = Chromosome))
```

Distribution of SNPs across chromosomes
```{r}
ggplot(data = maize_genotypes_positions) + geom_bar(mapping = aes(x = Chromosome)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
ggplot(data = teosinte_genotypes_positions) + geom_bar(mapping = aes(x = Chromosome)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```
Graphs of maize and teosinte are identical

### Missing data and heterozygosity

Create a new function called `is_heterozygous` that returns:
  1. TRUE for heterozygous positions
  2. FALSE for homozygous positions
  3. NA for missing data
```{r}
#function to check for heterozygosity
is_heterozygous <- function(genotype) {
  if(genotype == 'A/A'|
     genotype == 'G/G'|
     genotype == 'C/C'|
     genotype == 'T/T'){
    return(FALSE) #is homozygous
  }
  else if(genotype == '?/?'){
    return(NA) #is missing data
  }
  else{
    return(TRUE) #is heterozygous
  }
}
```


```{r}
#Preallocate new objects to store heterozygosity data
maize_heterozygosity <- 
  data.frame(matrix(nrow=nrow(maize_genotypes_positions) , ncol=length(maize_genotypes_positions)))
maize_heterozygosity[,1:3] <- maize_genotypes_positions[,1:3]
  

teosinte_heterozygosity <- 
  data.frame(matrix(nrow=nrow(teosinte_genotypes_positions) , ncol=length(teosinte_genotypes_positions)))
teosinte_heterozygosity[,1:3] <- teosinte_genotypes_positions[,1:3]

#Loop through maize genotypes file, determine heterozygosity for each element
for (i in 1:nrow(maize_genotypes_positions)) {
  for (j in 4:length(maize_genotypes_positions)) {
    maize_heterozygosity [i,j] <- is_heterozygous(maize_genotypes_positions[i,j])
  }
}

#Loop through teosinte genotypes file, determine heterozygosity for each element
for (i in 1:nrow(teosinte_genotypes_positions)) {
  for (j in 4:length(teosinte_genotypes_positions)) {
    teosinte_heterozygosity [i,j] <- is_heterozygous(teosinte_genotypes_positions[i,j])
  }
}

#calculate and store summary of maize heterozygosity:
maize_hetero_summary <- data.frame(3,2)
colnames(maize_hetero_summary) <- c('names', 'proportion')
maize_hetero_summary[1,1] <- 'prop_homozygous'
maize_hetero_summary[2,1] <- 'prop_heterozygous'
maize_hetero_summary[3,1] <- 'prop_missing'

maize_hetero_summary[1,2] <- 
  sum(maize_heterozygosity[,-1:-3] == FALSE, na.rm = TRUE) / (length(maize_heterozygosity[,-1:-3]) * nrow(maize_heterozygosity)) # number of homozygous/number elements
maize_hetero_summary[2,2] <- 
  sum(maize_heterozygosity[,-1:-3], na.rm = TRUE) / (length(maize_heterozygosity[,-1:-3]) * nrow(maize_heterozygosity)) #number of heterozygous/number elements
maize_hetero_summary[3,2] <- 
  sum(is.na(maize_heterozygosity[,-1,-3]), na.rm=TRUE) / (length(maize_heterozygosity[,-1:-3]) * nrow(maize_heterozygosity)) #number of missing/number elements

ggplot(data = maize_hetero_summary) + geom_col(mapping = aes(x = names, y = proportion)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

#calculate and store summary of teosinte heterozygosity:
teosinte_hetero_summary <- data.frame(3,2)
colnames(teosinte_hetero_summary) <- c('names', 'proportion')
teosinte_hetero_summary[1,1] <- 'prop_homozygous'
teosinte_hetero_summary[2,1] <- 'prop_heterozygous'
teosinte_hetero_summary[3,1] <- 'prop_missing'

teosinte_hetero_summary[1,2] <- 
  sum(teosinte_heterozygosity[,-1:-3] == FALSE, na.rm = TRUE) / (length(teosinte_heterozygosity[,-1:-3]) * nrow(teosinte_heterozygosity)) # number of homozygous/number elements
teosinte_hetero_summary[2,2] <- 
  sum(teosinte_heterozygosity[,-1:-3], na.rm = TRUE) / (length(teosinte_heterozygosity[,-1:-3]) * nrow(teosinte_heterozygosity)) #number of heterozygous/number elements
teosinte_hetero_summary[3,2] <- 
  sum(is.na(teosinte_heterozygosity[,-1,-3]), na.rm=TRUE) / (length(teosinte_heterozygosity[,-1:-3]) * nrow(teosinte_heterozygosity)) #number of missing/number elements

ggplot(data = teosinte_hetero_summary) + geom_col(mapping = aes(x = names, y = proportion)) + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))

```

### My own visualization
```{r}
#still working on this

```