---
title: "R_assignment_testing"
author: "David Hall"
date: "2023-03-12"
output: html_document
---

# R Assignment

load necessary R packages:
```{r packages}
library(tidyverse)
library(tidyr)
```

load data files `fang_et_al_genotypes.txt` and `snp_position.txt`:
```{r data}
fang_et_al_genotypes <- 
  read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2023/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt")

snp_position <- 
  read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2023/main/assignments/UNIX_Assignment/snp_position.txt")
```

## Data inspection
### Properties of `fang_et_al_genotypes`
### Properties of `snp_position`
## Data processing

### Maize data

Extract maize data, transpose, and convert SNP_ID from row names to a column:
```{r sort_transpose}
maize_genotypes <- filter(fang_et_al_genotypes, Group=='ZMMIL'|Group =='ZMMLR'|Group =='ZMMMR') %>% t() %>% as.data.frame() %>% rownames_to_column()
names(maize_genotypes)[1] <- "SNP_ID"
```

Join SNP_ID, chromosome, and position fields of `snp_position` with genotype data:
```{r join}
maize_genotypes_positions <- c(snp_position["SNP_ID"],snp_position["Chromosome"],snp_position["Position"]) %>% as.data.frame() %>% left_join(maize_genotypes, by = "SNP_ID")
```

sort by chromosome:
```{r chromosomes}
for (i in 1:10) {
  
  #filter by chromosome and arrange by increasing position value
  temp_object1 <- filter(maize_genotypes_positions, Chromosome == i & Position != 'multiple' & Position != 'unknown') %>% arrange(as.numeric(Position))
  
  #arrange by decreasing position value and replace '?' with '-'
  temp_object2 <- temp_object1 %>% arrange(desc(as.numeric(Position))) %>% mutate_if(is.character, str_replace_all, pattern = '[?]', replacement = '-')
  
  #assign stored data to object named by chromosome
  assign(paste0('maize_incr_chr_', i), temp_object1)
  assign(paste0('maize_decr_chr_', i), temp_object2)
}
```


### Teosinte data

Extract teosinte data, transpose, and convert SNP_ID from row names to a column:
```{r sort_transpose}
teosinte_genotypes <- filter(fang_et_al_genotypes, Group =='ZMPBA'|Group =='ZMPIL'|Group =='ZMPJA') %>% t() %>% as.data.frame() %>% rownames_to_column()
names(teosinte_genotypes)[1] <- "SNP_ID"
```

Join chromosome and position fields of `snp_position` with genotype data:
```{r join}
teosinte_genotypes_positions <- c(snp_position["SNP_ID"],snp_position["Chromosome"],snp_position["Position"]) %>% as.data.frame() %>% left_join(teosinte_genotypes, by = "SNP_ID")
```

sort by chromosome:
```{r chromosomes}
for (i in 1:10) {
  
  #filter by chromosome and arrange by increasing position value
  temp_object1 <- filter(teosinte_genotypes_positions, Chromosome == i & Position != 'multiple' & Position != 'unknown') %>% arrange(as.numeric(Position))
  #arrange by decreasing position value and replace '?' with '-'
  temp_object2 <- temp_object1 %>% arrange(desc(as.numeric(Position))) %>% mutate_if(is.character, str_replace_all, pattern = '[?]', replacement = '-')
  
  #assign stored data to object named by chromosome
  assign(paste0('teosinte_incr_chr_', i), temp_object1)
  assign(paste0('teosinte_decr_chr_', i), temp_object2)
}
```

## Visualization

### SNPs per chromosome

Distribution of SNPs on each chromosome
```{r}

ggplot(data = maize_genotypes_positions) + geom_point(aes(x = as.numeric(Position), y = Chromosome), color = 'blue')

ggplot(data = teosinte_genotypes_positions) + geom_point(aes(x = as.numeric(Position), y = Chromosome), color = 'red')
```

Distribution of SNPs across chromosomes
```{r}
ggplot(data = maize_genotypes_positions) + geom_bar(mapping = aes(x = Chromosome), fill = 'blue') + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
ggplot(data = teosinte_genotypes_positions) + geom_bar(mapping = aes(x = Chromosome), fill= 'red') + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```
Graphs of maize and teosinte are identical

### Missing data and heterozygosity

Create a new function called `is_heterozygous` that returns:
  1. TRUE for heterozygous positions
  2. FALSE for homozygous positions
  3. NA for missing data
```{r}
#function to check for heterozygosity
is_heterozygous <- function(genotype){
  return(ifelse(genotype == 'A/A' | genotype == 'T/T' | genotype == 'G/G' | genotype == 'C/C', FALSE, ifelse(genotype == '?/?', NA, TRUE)))
}
```

Use this function to determine heterozygosity at each site:
```{r}
#Preallocate new objects to store maize heterozygosity data
maize_heterozygosity <- data.frame(matrix(nrow=nrow(maize_genotypes_positions) , ncol=length(maize_genotypes_positions)))
maize_heterozygosity[,1:3] <- c(maize_genotypes_positions[,1:3]) #copy position and chromosome data
names(maize_heterozygosity) <- names(maize_genotypes_positions)

#check each element for heterozygosity and record it in the new data frame
maize_heterozygosity[,-1:-3] <- is_heterozygous(maize_genotypes_positions[,-1:-3])

##Preallocate new objects to store maize heterozygosity data
teosinte_heterozygosity <- data.frame(matrix(nrow=nrow(teosinte_genotypes_positions) , ncol=length(teosinte_genotypes_positions)))
teosinte_heterozygosity[,1:3] <- teosinte_genotypes_positions[,1:3] #copy position and chromosome data
names(teosinte_heterozygosity) <- names(teosinte_genotypes_positions)

#check each element for heterozygosity and record it in the new data frame
teosinte_heterozygosity[,-1:-3] <- is_heterozygous(teosinte_genotypes_positions[,-1:-3])
```

Summarize this heterozygosity data visually:
```{r}
#calculate the number of heterozygous, homozygous, and missing elements for each maize SNP (row)
j <- length(maize_heterozygosity)
maize_heterozygosity <- maize_heterozygosity %>%  
  mutate(heterozygous=rowSums(maize_heterozygosity[,4:j], na.rm=TRUE), #count TRUE values (heterozygous)
         homozygous=rowSums(maize_heterozygosity[,4:j]==FALSE,na.rm=TRUE), #switch TRUE values to homozygous and count
         missing=rowSums(is.na(maize_heterozygosity[,4:j]),na.rm=TRUE)) #switch TRUE values to missing and count

#calculate the number of heterozygous, homozygous, and missing elements for each teosinte SNP (row)
k <- length(teosinte_heterozygosity)
teosinte_heterozygosity <- teosinte_heterozygosity %>%  
  mutate(heterozygous=rowSums(teosinte_heterozygosity[,4:k], na.rm=TRUE), #count TRUE values (heterozygous)
         homozygous=rowSums(teosinte_heterozygosity[,4:k]==FALSE,na.rm=TRUE), #switch TRUE values to homozygous and count
         missing=rowSums(is.na(teosinte_heterozygosity[,4:k]),na.rm=TRUE)) #switch TRUE values to missing and count

#create summary table of maize and teosinte heterozygosity
hetero_summary <- data.frame(matrix(nrow=6,ncol=3))
names(hetero_summary) <- c('group','status','proportion') 
hetero_summary[,'group'] <- c('maize','maize','maize','teosinte','teosinte','teosinte')
hetero_summary[,'status'] <- c('missing','heterozygous','homozygous')
hetero_summary[1,'proportion'] <- sum(maize_heterozygosity[,'missing']) / (nrow(maize_heterozygosity)*(j-6)) #missing proportion maize
hetero_summary[2,'proportion'] <- sum(maize_heterozygosity[,'heterozygous']) / (nrow(maize_heterozygosity)*(j-6)) #heterozygous proportion maize
hetero_summary[3,'proportion'] <- sum(maize_heterozygosity[,'homozygous']) / (nrow(maize_heterozygosity)*(j-6)) #homozygous proportion maize
hetero_summary[4,'proportion'] <- sum(teosinte_heterozygosity[,'missing']) / (nrow(teosinte_heterozygosity)*(k-6)) #missing proportion teosinte
hetero_summary[5,'proportion'] <- sum(teosinte_heterozygosity[,'heterozygous']) / (nrow(teosinte_heterozygosity)*(k-6)) #heterozygous proportion teosinte
hetero_summary[6,'proportion'] <- sum(teosinte_heterozygosity[,'homozygous']) / (nrow(teosinte_heterozygosity)*(k-6)) #homozygous proportion teosinte

#Plot the data
ggplot(data = hetero_summary) + geom_col(mapping = aes(x = group, y = proportion, fill= status), position = 'stack') + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
```
There are a higher proportion of homozygous SNPs and a lower proportion of heterozygous SNPs in maize compared to teosinte.
Proportion of missing data is about the same.

### My own visualization
```{r}
#still working on this

```