---
title: "R Assignment"
author: "David Hall"
date: "2023-03-12"
output:
  html_document: default
  pdf_document: default
---

load necessary R packages:
```{r packages, message=FALSE}
library(tidyverse)
library(tidyr)
```

load data files `fang_et_al_genotypes.txt` and `snp_position.txt`:
```{r data, message=FALSE}
fang_et_al_genotypes <- 
  read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2023/main/assignments/UNIX_Assignment/fang_et_al_genotypes.txt")

snp_position <- 
  read_tsv("https://raw.githubusercontent.com/EEOB-BioData/BCB546_Spring2023/main/assignments/UNIX_Assignment/snp_position.txt")
```

## Data inspection

### Properties of `fang_et_al_genotypes`

Code used to inspect `fang_et_al_genotypes`:
```
print(object.size(fang_et_al_genotypes), units = 'Mb')
dim(fang_et_al_genotypes)
str(fang_et_al_genotypes)
view(fang_et_al_genotypes)
```

from this I learned:

1. fang_et_al_genotypes is 22.1 Mb in size
2. fang_et_al_genotypes has 2782 rows and 986 columns
3. All columns contain values recognized as character type
4. The file has the following structure: Sample_ID, JG_OTU, Group, and then genoytpe data for each SNP_ID

### Properties of `snp_position`

Code used to inspect `snp_position`:
```
print(object.size(snp_position), units = 'Kb')
dim(snp_position)
str(snp_position)
view(snp_position)
```

from this I learned:

1. snp_position is 351 Kb in size
2. snp_position has 915 rows and 15 columns
3. The columns contain values recognized as character or numeric type
4. The columns we need (SNP_ID, Chromosome, Position) are all of type character, even though one might expect that Chromosome and Position would be numeric 


## Data processing


### Maize data


Extract maize data, transpose, and convert SNP_ID from row names to a column:
```{r maize_sort_transpose}
maize_genotypes <- filter(fang_et_al_genotypes, Group=='ZMMIL' | Group =='ZMMLR' | Group =='ZMMMR') %>% t() %>% as.data.frame() %>% rownames_to_column()
names(maize_genotypes)[1] <- 'SNP_ID' 
```

Join SNP_ID, chromosome, and position fields of `snp_position` with genotype data:
```{r maize_join}
maize_genotypes_positions <- c(snp_position['SNP_ID'], snp_position['Chromosome'], snp_position['Position']) %>% as.data.frame() %>% left_join(maize_genotypes, by='SNP_ID')
```

sort by chromosome:
```{r maize_chromosomes}
for (i in 1:10) {
  
  #filter by chromosome and arrange by increasing position value
  temp_object1 <- filter(maize_genotypes_positions, Chromosome==i & Position!='multiple' & Position!='unknown') %>% arrange(as.numeric(Position))
  
  #arrange by decreasing position value and replace '?' with '-'
  temp_object2 <- temp_object1 %>% arrange(desc(as.numeric(Position))) %>% mutate_if(is.character, str_replace_all, pattern='[?]', replacement='-')
  
  #assign stored data to object named by chromosome
  write.table(temp_object1, file=paste0('maize_incr_chr_', i, '.txt'), quote=FALSE, sep='\t', row.names=FALSE, col.names=TRUE)
  write.table(temp_object2, file=paste0('maize_decr_chr_', i, '.txt'), quote=FALSE, sep='\t', row.names=FALSE, col.names=TRUE)
}
```


### Teosinte data


Extract teosinte data, transpose, and convert SNP_ID from row names to a column:
```{r teosinte_sort_transpose}
teosinte_genotypes <- filter(fang_et_al_genotypes, Group =='ZMPBA' | Group =='ZMPIL' | Group =='ZMPJA') %>% t() %>% as.data.frame() %>% rownames_to_column()
names(teosinte_genotypes)[1] <- 'SNP_ID' 
```

Join chromosome and position fields of `snp_position` with genotype data:
```{r teosinte_join}
teosinte_genotypes_positions <- c(snp_position['SNP_ID'], snp_position['Chromosome'], snp_position['Position']) %>% as.data.frame() %>% left_join(teosinte_genotypes, by='SNP_ID')
```

sort by chromosome:
```{r teosinte_chromosomes}
for (i in 1:10) {
  
  #filter by chromosome and arrange by increasing position value
  temp_object1 <- filter(teosinte_genotypes_positions, Chromosome==i & Position!='multiple' & Position!='unknown') %>% arrange(as.numeric(Position))
  #arrange by decreasing position value and replace '?' with '-'
  temp_object2 <- temp_object1 %>% arrange(desc(as.numeric(Position))) %>% mutate_if(is.character, str_replace_all, pattern='[?]', replacement='-')
  
  #assign stored data to object named by chromosome
  write.table(temp_object1, file=paste0('teosinte_incr_chr_', i, '.txt'), quote=FALSE, sep='\t', row.names=FALSE, col.names=TRUE)
  write.table(temp_object2, file=paste0('teosinte_decr_chr_', i, '.txt'), quote=FALSE, sep='\t', row.names=FALSE, col.names=TRUE)
}
```


## Visualization


### SNPs per chromosome


Distribution of SNPs on each chromosome
```{r SNPs_across_chrms}
#Plot maize SNPs
ggplot(data = maize_genotypes_positions %>% group_by(Chromosome) %>% filter(Chromosome!='multiple') %>%  filter(Chromosome!='unknown') %>% filter(Position!='multiple') %>% filter(Position!='unknown')) + #remove multiple and unknown positions
  geom_point(aes(x = as.integer(Position), y = as.integer(Chromosome)), color = 'forestgreen') + 
  labs(x='Position (bases)', y='Chromosome', title='Maize: SNP distrubution by chromosome') + 
  scale_y_continuous(breaks = c(1:10))

#Plot teosinte SNPs
ggplot(data = teosinte_genotypes_positions %>% group_by(Chromosome) %>% filter(Chromosome!='multiple') %>%  filter(Chromosome!='unknown') %>% filter(Position!='multiple') %>% filter(Position!='unknown')) +  #remove multiple and unknown positions
  geom_point(aes(x = as.integer(Position), y = as.integer(Chromosome)), color = 'firebrick1') + 
  labs(x='Position (bases)', y='Chromosome', title='Teosinte: SNP distrubution by chromosome') + 
  scale_y_continuous(breaks = c(1:10))
```


Number of SNPs per chromosome:
```{r SNPs_per_chrm}
#Plot maize SNPs per chromosome
ggplot(data = maize_genotypes_positions) + 
  geom_bar(mapping=aes(x=Chromosome), fill='forestgreen') + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + 
  scale_x_discrete(limits=c('1','2','3','4','5','6','7','8','9','10','multiple','unknown')) + 
  labs(x='Chromosome', y='Number of SNPs', title='Maize: SNPs per Chromosome')

#Plot teosinte SNPs per chromosome
ggplot(data = teosinte_genotypes_positions) + 
  geom_bar(mapping=aes(x=Chromosome), fill='firebrick1') + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + 
  scale_x_discrete(limits=c('1','2','3','4','5','6','7','8','9','10','multiple','unknown')) + 
  labs(x='Chromosome', y='Number of SNPs', title='Teosinte: SNPs per Chromosome')
```


Graphs of maize and teosinte are identical


### Missing data and heterozygosity


Create a new function called `zygosity` that returns:

  1. 'heterozygous' for heterozygous positions
  2. 'homozygous' for homozygous positions
  3. 'missing' for missing data
  
```{r zygosity_fun}
#function to check for heterozygosity
zygosity <- function(genotype){
  return(ifelse(genotype=='A/A' | genotype=='T/T' | genotype=='G/G' | genotype=='C/C', 'homozygous', ifelse(genotype=='?/?', 'missing', 'heterozygous')))
}
```

Plot the heterozygosity data for each group in `fang_et_al_genotypes`
```{r zygosity_group}
#make the fang_et_al_genotypes long (tidy the data)
fang_genotypes_long <- pivot_longer(fang_et_al_genotypes, -1:-3, names_to ='SNP_ID', values_to ='genotype')

#determine zygosity of genotypes and store in a new column
fang_genotypes_long <- mutate(fang_genotypes_long, zygosity=zygosity(genotype))

#plot zygosity by group
fang_genotypes_long <- fang_genotypes_long %>% mutate(zygosity=fct_relevel(zygosity, 'missing','homozygous','heterozygous'))#'missing' is plotted on top
ggplot(data = fang_genotypes_long) + 
  geom_bar(mapping = aes(x=Group, fill=zygosity), position='fill') + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + 
  labs(x='Group', y='Proportion of SNPs', title='Zygosity across groups') 
```

For maize:

  - ZMMIL and ZMMMR groups have very little heterozygosity
  - ZMMLR group has relatively higher heterozygosity, and lower proportions of missing genotype data
  
For Teosinte:

  - ZMPBA has the highest heterozygosity of any group
  - ZMPJA also has relatively high proportions of heterozygosity, similar to the maize group ZMMLR
  - ZMPIL has low levels of heterozygosity, similar to the maize groups ZMMIL and ZMMMR
  - Proportion of missing data is comparable between the three teosinte groups
  
  
To dissect this further, plot the heterozygosity data by sample within each maize and teosinte group:
```{r fang_longer_zygosity}
#Add a column that numbers samples
fang_numbers <- mutate(fang_et_al_genotypes, Sample_number = 1:nrow(fang_et_al_genotypes),.before=1)

#make the data long and add zygosity column
fang_numbers_long <- pivot_longer(fang_numbers, -1:-4, names_to = 'SNP_ID', values_to = 'genotype')
fang_numbers_long <- mutate(fang_numbers_long, zygosity=zygosity(genotype))

#change the order so missing is plotted on top
fang_numbers_long <- fang_numbers_long %>% mutate(zygosity = fct_relevel(zygosity, 'missing','homozygous','heterozygous'))
```

plot maize zygosity based on samples within groups:
```{r maize_zygosity_sample}
#ZMMIL
ggplot(data = fang_numbers_long %>% filter(Group=='ZMMIL')) + 
  geom_histogram(mapping = aes(x=Sample_number, fill=zygosity), binwidth=1, position='fill') + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + 
  labs(x='sample number', y='proportion of SNPs', title='Maize: Zygosity of ZMMIL samples')

#ZMMLR
ggplot(data=fang_numbers_long %>% filter(Group=='ZMMLR')) + 
  geom_histogram(mapping=aes(x=Sample_number, fill=zygosity), binwidth=1, position='fill') + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + 
  labs(x='sample number', y='proportion of SNPs', title='Maize: Zygosity of ZMMLR samples')

#ZMMMR
ggplot(data=fang_numbers_long %>% filter(Group=='ZMMMR')) + 
  geom_histogram(mapping=aes(x=Sample_number, fill=zygosity), binwidth=1, position='fill') + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + 
  labs(x='sample number', y='proportion of SNPs', title='Maize: Zygosity of ZMMMR samples')
```


Samples vary group to group as noted above, but are relatively consist across samples within a group. small sample size for ZMMMR compared to the other two groups.


Plot teosinte zygosity based on samples within groups:
```{r teosinte_zygosity_sample}
#ZMPBA
ggplot(data = fang_numbers_long %>% filter(Group=='ZMPBA')) + 
  geom_histogram(mapping=aes(x=Sample_number, fill=zygosity), binwidth=1, position='fill') + 
  theme(axis.text.x=element_text(angle =90, vjust=0.5, hjust=1)) + 
  labs(x='sample number', y='proportion of SNPs', title='Teosinte: Zygosity of ZMPBA samples')

#ZMPIL
ggplot(data=fang_numbers_long %>% filter(Group=='ZMPIL')) + 
  geom_histogram(mapping=aes(x=Sample_number, fill=zygosity), binwidth=1, position ='fill') + 
  theme(axis.text.x = element_text(angle=90, vjust=0.5, hjust=1)) + 
  labs(x='sample number', y='proportion of SNPs', title='Teosinte: Zygosity of ZMPIL samples')

#ZMPJA
ggplot(data=fang_numbers_long %>% filter(Group=='ZMPJA')) + 
  geom_histogram(mapping=aes(x=Sample_number, fill=zygosity), binwidth=1, position='fill') + 
  theme(axis.text.x = element_text(angle=90, vjust=0.5, hjust=1)) + 
  labs(x='sample number', y='proportion of SNPs', title='Teosinte: Zygosity of ZMPJA samples')
```

Like in maize, heterozygosity varies group to group but is relatively consistent across samples within a group.


### My own visualization: heterozygosity across chromosomes and species
```{r zygosity_per_chrm}
#make the data long and add a new column with zygosity data
maize_genotypes_long <- pivot_longer(maize_genotypes_positions, -1:-3, names_to='Sample', values_to='genotype')
maize_genotypes_long <- mutate(maize_genotypes_long, zygosity = zygosity(genotype))
teosinte_genotypes_long <- pivot_longer(teosinte_genotypes_positions, -1:-3, names_to='sample', values_to='genotype')
teosinte_genotypes_long <- mutate(teosinte_genotypes_long, zygosity = zygosity(genotype))

#Plot Maize 
maize_genotypes_long <- maize_genotypes_long %>% mutate(zygosity = fct_relevel(zygosity, 'missing','homozygous','heterozygous'))#Plot 'missing' on top
ggplot(data=maize_genotypes_long) + 
  geom_bar(mapping=aes(x=Chromosome ,fill=zygosity), position ='fill') + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + 
  labs(x='Chromosome', y='Proportion of SNPs', title='Maize: Zygosity across chromosomes')

#Plot Teosinte
teosinte_genotypes_long <- teosinte_genotypes_long %>% mutate(zygosity = fct_relevel(zygosity, 'missing','homozygous','heterozygous'))#Plot 'missing' on top
ggplot(data=teosinte_genotypes_long) + 
  geom_bar(mapping=aes(x=Chromosome ,fill=zygosity), position='fill') + 
  theme(axis.text.x=element_text(angle=90, vjust=0.5, hjust=1)) + 
  labs(x='Chromosome', y='Proportion of SNPs', title='Teosinte: Zygosity across chromosomes')
```

heterozygosity does not vary much across chromosomes. Across species, maize has a higher proportion of homozygous SNP postitions compared to teosinte. However, as seen above, heterozygosity varies quite significantly from group to group within both species.


